### DISCUSSION TITLE
Inserting variables into topology XML

### DISCUSSION THREAD

**Discussion by teddylambert** [nasa/fprime#1234] [Q&A]
Note: We are using fprime v2.0.1 for the project

I am working on a project where, for a variety of reasons, we are using the port number a message on an N-width bus comes in on to tack some header information on to the message. Because of this, I wanted to maintain the port numbers for specific connections in `AcConstants.ini` and use variables in the topology XML so that in the event a port number does change, just the AcConstants file changes instead of digging through the topology files.

I know this is possible with component XMLs, as seen with the static memory allocator:
```
<port name="bufferAllocate" data_type="Fw::BufferGet"  kind="guarded_input" max_number="$StaticMemoryAllocations">
</port>
```
and in `AcConstants.ini`:
`StaticMemoryAllocations = 4`

However, trying to do this in the topology XML, such as:
```
<connection name="ExampleCmdReg">
        <source component="comp1" port="CmdReg" type="CmdReg" num="0"/>
        <target component="comp2" port="portIn" type="Serial" num="$comp1RegInputPort"/>
</connection>
```
and in `AcConstants.ini`:
`comp1RegInputPort = 3`

generates the following error:
```
.../Top/TestTopologyAppAi.xml:66:0:ERROR:RELAXNGV:RELAXNG_ERR_INVALIDATTR: Invalid attribute num for element target
.../Top/TestTopologyAppAi.xml:64:0:ERROR:RELAXNGV:RELAXNG_ERR_INTERSEQ: Invalid sequence in interleave
.../Top/TestTopologyAppAi.xml:64:0:ERROR:RELAXNGV:RELAXNG_ERR_CONTENTVALID: Element connection failed to validate content
```

Is what I want to do possible? And if so, what do I need to do for the variables to be parsed correctly?

**Comment by LeStarch** [comment]
I don't know this is possible in the XML.  I believe it is possible in FPP (v3), although I have not tested it.  We've also had success using a templating language (jinja2, mustache, etc) to build the topology Ai.xml from a templatized version.

**Reply by SterlingPeet** [reply]
Yea, I would think that a ``cookiecutter`` or ``jinja`` template would be the easiest here, when the autocoder doesn't support the templating functionality.
